In [1]:
import numpy as np
import pandas as pd


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.1 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.3.1 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/traitlets/config/application.py", line 992, in launch_instance
    app.start()
  File "/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/ipykernel/kernelapp.py", line 711, in start
    self.io_

AttributeError: _ARRAY_API not found

In [2]:
pip install polars

Note: you may need to restart the kernel to use updated packages.


In [3]:
facebook_posts = pd.read_csv("2024_fb_posts_president_scored_anon.csv")

In [4]:
facebook_posts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19009 entries, 0 to 19008
Data columns (total 56 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Facebook_Id                                                       19009 non-null  object 
 1   post_id                                                           19009 non-null  object 
 2   Page Category                                                     16537 non-null  object 
 3   Page Admin Top Country                                            16280 non-null  object 
 4   Post Created                                                      19009 non-null  object 
 5   Post Created Date                                                 19009 non-null  object 
 6   Post Created Time                                                 19009 non-null  object 
 7   Type                           

In [5]:
#Removing columsn with 0 non null values: 

facebook_posts_clean = facebook_posts.drop(columns=['Sponsor Id', 'Sponsor Name', 'Sponsor Category', 'illuminating_scored_messageelection_integrity_Truth_illuminating'])

print("Done! Removed 4 empty columns.")

Done! Removed 4 empty columns.


In [6]:
#Cleaning data: 

# Simple data cleaning for Facebook posts dataset
import pandas as pd

# 1. Fix Total Interactions column (convert to numeric)
facebook_posts_clean['Total Interactions'] = pd.to_numeric(facebook_posts_clean['Total Interactions'], errors='coerce')

# 2. Convert date columns to datetime
facebook_posts_clean['Post Created Date'] = pd.to_datetime(facebook_posts_clean['Post Created Date'], errors='coerce')
facebook_posts_clean['Post Created Time'] = pd.to_datetime(facebook_posts_clean['Post Created Time'], errors='coerce')

print("Cleaning complete!")
print("Total Interactions now numeric:", facebook_posts_clean['Total Interactions'].dtype)
print("Date columns converted to datetime")

Cleaning complete!
Total Interactions now numeric: float64
Date columns converted to datetime


/var/folders/9c/9gm285k54tz9ysvymk9gtd400000gn/T/ipykernel_66736/1558518615.py:11: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  facebook_posts_clean['Post Created Time'] = pd.to_datetime(facebook_posts_clean['Post Created Time'], errors='coerce')


In [7]:
facebook_posts_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19009 entries, 0 to 19008
Data columns (total 52 columns):
 #   Column                                     Non-Null Count  Dtype         
---  ------                                     --------------  -----         
 0   Facebook_Id                                19009 non-null  object        
 1   post_id                                    19009 non-null  object        
 2   Page Category                              16537 non-null  object        
 3   Page Admin Top Country                     16280 non-null  object        
 4   Post Created                               19009 non-null  object        
 5   Post Created Date                          19009 non-null  datetime64[ns]
 6   Post Created Time                          19009 non-null  datetime64[ns]
 7   Type                                       16544 non-null  object        
 8   Total Interactions                         14398 non-null  float64       
 9   Likes            

In [8]:
#Finding out if any columns are actually dictionaries or lists: 

# Find and inspect object columns to see what's actually in them
import pandas as pd

# Get all object columns
object_columns = facebook_posts_clean.select_dtypes(include=['object']).columns

print("Object columns found:")
for col in object_columns:
    print(f"\n--- {col} ---")
    print(f"Sample values:")
    # Show first 3 non-null values
    sample_values = facebook_posts_clean[col].dropna().head(3).tolist()
    for i, val in enumerate(sample_values):
        print(f"  {i+1}: {val} (type: {type(val).__name__})")
    
    # Check if any values are lists or dicts
    has_lists = any(isinstance(val, list) for val in facebook_posts_clean[col].dropna().head(10))
    has_dicts = any(isinstance(val, dict) for val in facebook_posts_clean[col].dropna().head(10))
    
    if has_lists:
        print(f"  ⚠️  Contains LISTS")
    if has_dicts:
        print(f"  ⚠️  Contains DICTIONARIES")
    if not has_lists and not has_dicts:
        print(f"  ✅ Just regular strings")

Object columns found:

--- Facebook_Id ---
Sample values:
  1: a6cb7db6850459b954f7272e14d770617022639f8847cf25e6859b9453a03813 (type: str)
  2: a6cb7db6850459b954f7272e14d770617022639f8847cf25e6859b9453a03813 (type: str)
  3: a6cb7db6850459b954f7272e14d770617022639f8847cf25e6859b9453a03813 (type: str)
  ✅ Just regular strings

--- post_id ---
Sample values:
  1: 8570b69695e00d8f06b12398ed525497e1712b5369c6fc2138fe98f69811c138 (type: str)
  2: 41ec27cecd8af40007a9faf8c3e5c9225bcff0b8d58856563bbc9d59d7dbb6e6 (type: str)
  3: 1dcb5e00cd1c8d7ee141922f50f29e59e96328231b6937870e87029c97b26377 (type: str)
  ✅ Just regular strings

--- Page Category ---
Sample values:
  1: PERSON (type: str)
  2: PERSON (type: str)
  3: PERSON (type: str)
  ✅ Just regular strings

--- Page Admin Top Country ---
Sample values:
  1: US (type: str)
  2: US (type: str)
  3: US (type: str)
  ✅ Just regular strings

--- Post Created ---
Sample values:
  1: 2023-09-04 19:31:16 EDT (type: str)
  2: 2023-09-06 20:00:5

In [14]:
import pandas as pd
import time

# Script 2 Part 1: Pandas Overall Descriptive Statistics for Facebook Dataset
print("=== SCRIPT 2 PART 1: PANDAS OVERALL STATISTICS ===")

# Set pandas display options to show all columns and format numbers
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.float_format', '{:.3f}'.format)  # 3 decimal places
pd.set_option('display.precision', 3)  # 3 decimal precision

# Start timing
start_time = time.time()

# Define our column categories (specific to Facebook dataset)
numeric_columns = [
    'Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care',
    'Post Views', 'Total Views', 'Total Views For All Crossposts', 'Overperforming Score',
    'Total Interactions',  # After cleaning
    'advocacy_msg_type_illuminating', 'issue_msg_type_illuminating', 'attack_msg_type_illuminating',
    'image_msg_type_illuminating', 'cta_msg_type_illuminating', 'engagement_cta_subtype_illuminating',
    'fundraising_cta_subtype_illuminating', 'voting_cta_subtype_illuminating',
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'lgbtq_issues_topic_illuminating',
    'military_topic_illuminating', 'race_and_ethnicity_topic_illuminating', 'safety_topic_illuminating',
    'social_and_cultural_topic_illuminating', 'technology_and_privacy_topic_illuminating',
    'womens_issue_topic_illuminating', 'incivility_illuminating', 'scam_illuminating',
    'freefair_illuminating', 'fraud_illuminating'
]

categorical_columns = [
    'Facebook_Id', 'post_id', 'Page Category', 'Page Admin Top Country', 
    'Post Created', 'Type', 'Video Share Status', 'Is Video Owner?', 'Video Length'
]

# Note: Post Created Date and Post Created Time were converted to datetime during cleaning
datetime_columns = ['Post Created Date', 'Post Created Time']

print("OVERALL DATASET STATISTICS")
print("=" * 60)

print(f"\nDataset Info:")
print(f"Total Rows: {len(facebook_posts_clean)}")
print(f"Total Columns: {len(facebook_posts_clean.columns)}")
print(f"Numeric Columns: {len(numeric_columns)}")
print(f"Categorical Columns: {len(categorical_columns)}")

# NUMERIC STATISTICS
print("\n" + "=" * 60)
print("NUMERIC COLUMNS STATISTICS")
print("=" * 60)

# Get numeric statistics using describe()
numeric_stats = facebook_posts_clean[numeric_columns].describe()
print("\nPandas .describe() output:")
print(numeric_stats)

# Add additional stats not in describe()
print("\nAdditional Numeric Statistics:")
additional_stats = facebook_posts_clean[numeric_columns].agg(['sum', 'var'])
print(additional_stats)

# CATEGORICAL STATISTICS
print("\n" + "=" * 60)
print("CATEGORICAL COLUMNS STATISTICS")
print("=" * 60)

categorical_summary = {}
for col in categorical_columns:
    print(f"\n--- {col} ---")
    
    # Basic stats
    count = facebook_posts_clean[col].count()
    unique_count = facebook_posts_clean[col].nunique()
    
    # Most frequent value
    if not facebook_posts_clean[col].mode().empty:
        mode_value = facebook_posts_clean[col].mode().iloc[0]
        mode_count = facebook_posts_clean[col].value_counts().iloc[0]
    else:
        mode_value = 'N/A'
        mode_count = 0
    
    print(f"Count: {count}")
    print(f"Unique Values: {unique_count}")
    print(f"Most Frequent: '{mode_value}' (appears {mode_count} times)")
    
    # Top 5 values
    print("Top 5 Values:")
    top_values = facebook_posts_clean[col].value_counts().head(5)
    for i, (val, count_val) in enumerate(top_values.items(), 1):
        print(f"  {i}. '{val}': {count_val}")
    
    # Store for summary
    categorical_summary[col] = {
        'count': count,
        'unique': unique_count,
        'mode': mode_value,
        'mode_count': mode_count
    }

# End timing
end_time = time.time()
execution_time = end_time - start_time

# DATETIME STATISTICS
print("\n" + "=" * 60)
print("DATETIME COLUMNS STATISTICS")
print("=" * 60)

for col in datetime_columns:
    print(f"\n--- {col} ---")
    if col in facebook_posts_clean.columns:
        print(f"Count: {facebook_posts_clean[col].count()}")
        print(f"Earliest: {facebook_posts_clean[col].min()}")
        print(f"Latest: {facebook_posts_clean[col].max()}")
        print(f"Date Range: {facebook_posts_clean[col].max() - facebook_posts_clean[col].min()}")
    else:
        print(f"Column '{col}' not found in dataset")

print("\n" + "=" * 60)
print("EXECUTION SUMMARY")
print("=" * 60)
print(f"Pandas Part 1 execution time: {execution_time:.4f} seconds")
print(f"Dataset shape: {facebook_posts_clean.shape}")
print(f"Numeric columns analyzed: {len(numeric_columns)}")
print(f"Categorical columns analyzed: {len(categorical_columns)}")
print(f"Datetime columns analyzed: {len(datetime_columns)}")

# Save results for comparison with other scripts
pandas_results_part1 = {
    'numeric_stats': numeric_stats,
    'additional_stats': additional_stats,
    'categorical_summary': categorical_summary,
    'execution_time': execution_time,
    'dataset_shape': facebook_posts_clean.shape
}

print(f"\nResults saved in 'pandas_results_part1' variable for comparison")

=== SCRIPT 2 PART 1: PANDAS OVERALL STATISTICS ===
OVERALL DATASET STATISTICS

Dataset Info:
Total Rows: 19009
Total Columns: 52
Numeric Columns: 41
Categorical Columns: 9

NUMERIC COLUMNS STATISTICS

Pandas .describe() output:
           Likes  Comments    Shares       Love       Wow      Haha       Sad  \
count  19009.000 19009.000 19009.000  19009.000 19009.000 19009.000 19009.000   
mean    2377.695   901.583   320.539    413.877     5.868   105.720    10.172   
std    11253.470  3681.980  1722.160   3730.940    52.947   942.034   418.325   
min        0.000     0.000     0.000      0.000     0.000     0.000     0.000   
25%       31.000     8.000     3.000      0.000     0.000     0.000     0.000   
50%      139.000    48.000    21.000      4.000     1.000     2.000     0.000   
75%      738.000   354.000   102.000     70.000     3.000    29.000     2.000   
max   351979.000 93872.000 76150.000 244482.000  4345.000 99276.000 56111.000   

          Angry      Care  Post Views  Tot

In [15]:
### GROUPING BY AND AGGREGATIONS: 

import pandas as pd
import time

# Script 2 Part 2: Pandas Groupby Analysis for Facebook Dataset
print("=== SCRIPT 2 PART 2: PANDAS GROUPBY ANALYSIS ===")

# Set pandas display options
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)
pd.set_option('display.float_format', '{:.3f}'.format)
pd.set_option('display.precision', 3)

# Start timing
start_time = time.time()

# Key columns for groupby analysis
engagement_columns = ['Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care']
view_columns = ['Post Views', 'Total Views', 'Total Interactions']
topic_columns = [
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'safety_topic_illuminating'
]

print("=" * 70)
print("1. GROUPED BY FACEBOOK_ID (Page Analysis)")
print("=" * 70)

# Group by Facebook_Id
grouped_by_page = facebook_posts_clean.groupby('Facebook_Id')

print("\nPage Activity Summary:")
page_activity = grouped_by_page.agg({
    'post_id': 'count',  # Number of posts per page
    'Likes': ['sum', 'mean', 'max'],
    'Comments': ['sum', 'mean', 'max'],
    'Shares': ['sum', 'mean', 'max'],
    'Total Interactions': ['sum', 'mean', 'max']
}).round(3)

# Flatten column names
page_activity.columns = ['_'.join(col).strip() for col in page_activity.columns]
print(page_activity.head(10))

print("\nTop 5 Pages by Total Likes:")
top_likes = grouped_by_page['Likes'].sum().sort_values(ascending=False).head(5)
print(top_likes)

print("\nTop 5 Pages by Total Comments:")
top_comments = grouped_by_page['Comments'].sum().sort_values(ascending=False).head(5)
print(top_comments)

print("\nTop 5 Pages by Post Count:")
top_posts = grouped_by_page.size().sort_values(ascending=False).head(5)
print(top_posts)

print("\nTopic Coverage by Page (Top 5 pages for each topic):")
for topic in topic_columns[:3]:  # Show first 3 topics
    topic_by_page = grouped_by_page[topic].sum().sort_values(ascending=False).head(5)
    print(f"\n{topic}:")
    print(topic_by_page)

print("\n" + "=" * 70)
print("2. GROUPED BY FACEBOOK_ID + PAGE CATEGORY")
print("=" * 70)

# Group by Facebook_Id and Page Category
grouped_by_page_category = facebook_posts_clean.groupby(['Facebook_Id', 'Page Category'])

print("\nPage + Category Activity Summary:")
page_cat_activity = grouped_by_page_category.agg({
    'post_id': 'count',
    'Likes': ['sum', 'mean'],
    'Comments': ['sum', 'mean'],
    'Shares': ['sum', 'mean'],
    'Total Interactions': ['sum', 'mean']
}).round(3)

# Flatten column names
page_cat_activity.columns = ['_'.join(col).strip() for col in page_cat_activity.columns]
print(page_cat_activity.head(15))

print("\nAverage Engagement by Page Category:")
category_stats = facebook_posts_clean.groupby('Page Category').agg({
    'Likes': ['count', 'mean', 'sum'],
    'Comments': ['count', 'mean', 'sum'],
    'Shares': ['count', 'mean', 'sum'],
    'Total Interactions': ['count', 'mean', 'sum']
}).round(3)

# Flatten column names
category_stats.columns = ['_'.join(col).strip() for col in category_stats.columns]
print(category_stats)

print("\nPost Type Distribution by Page Category:")
post_type_by_category = pd.crosstab(facebook_posts_clean['Page Category'], 
                                   facebook_posts_clean['Type'], 
                                   margins=True)
print(post_type_by_category)

print("\nTopic Focus by Page Category:")
topic_by_category = facebook_posts_clean.groupby('Page Category')[topic_columns].mean().round(3)
print(topic_by_category)

# End timing
end_time = time.time()
execution_time = end_time - start_time

print("\n" + "=" * 70)
print("EXECUTION SUMMARY")
print("=" * 70)
print(f"Pandas Part 2 execution time: {execution_time:.3f} seconds")
print(f"Unique Facebook_Id values: {facebook_posts_clean['Facebook_Id'].nunique()}")
print(f"Unique Page Categories: {facebook_posts_clean['Page Category'].nunique()}")
print(f"Total page-category combinations: {len(page_cat_activity)}")

# Save results for comparison
pandas_results_part2 = {
    'page_activity': page_activity,
    'page_cat_activity': page_cat_activity,
    'category_stats': category_stats,
    'topic_by_category': topic_by_category,
    'execution_time': execution_time
}

print(f"\nResults saved in 'pandas_results_part2' variable for comparison")

=== SCRIPT 2 PART 2: PANDAS GROUPBY ANALYSIS ===
1. GROUPED BY FACEBOOK_ID (Page Analysis)

Page Activity Summary:
                                                                  post_id_count  \
Facebook_Id                                                                       
058bd86861262fa71733f8515c34b729d9ebcad3d318e2a133a6c572dd80ba0b            210   
278df0e37b61bf8a5a6374873b38faddd50fffea73eaeb55c3c6240f79d1ba9f             23   
32fc18da91029ff09bf74fe9887eace6b5d2145809d583f696e344530508b064           9013   
459abece909c3b58ecbea0ad3bf398ae90216c170fff442979f2ac3f72f4c190             22   
4bd0a74df1b4df50157a304ace4291c947560a05da059ea6c8935cc11de7014d             98   
7ec2cb4abf8effe3d91de57944c56b938b4f33059a6e33dfe23b549cf229989a            616   
7f5731ac8e6959576e781064b5fa76e059200cdfc3f4e0d7b68bc98792d2cf5b           1158   
834f691e4ca1c67a7e33db1638caa579b426082e25427c63cb98679fb36af452             23   
8e8dac0ee6f22255b9d9365845a7fb830627cc27d34934c4f097689

### Polars

In [9]:
pip install --upgrade numpy pyarrow polars

Note: you may need to restart the kernel to use updated packages.


In [10]:
import time
import math
from collections import Counter

# Script 1 Part 1: Pure Python Descriptive Statistics for Facebook Dataset
print("=== SCRIPT 1 PART 1: PURE PYTHON OVERALL STATISTICS ===")

# Start timing
start_time = time.time()

# Convert DataFrame to list of dictionaries for pure Python processing
facebook_data = facebook_posts_clean.to_dict('records')

# Define our column categories (same as Pandas)
numeric_columns = [
    'Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care',
    'Post Views', 'Total Views', 'Total Views For All Crossposts', 'Overperforming Score',
    'Total Interactions',
    'advocacy_msg_type_illuminating', 'issue_msg_type_illuminating', 'attack_msg_type_illuminating',
    'image_msg_type_illuminating', 'cta_msg_type_illuminating', 'engagement_cta_subtype_illuminating',
    'fundraising_cta_subtype_illuminating', 'voting_cta_subtype_illuminating',
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'lgbtq_issues_topic_illuminating',
    'military_topic_illuminating', 'race_and_ethnicity_topic_illuminating', 'safety_topic_illuminating',
    'social_and_cultural_topic_illuminating', 'technology_and_privacy_topic_illuminating',
    'womens_issue_topic_illuminating', 'incivility_illuminating', 'scam_illuminating',
    'freefair_illuminating', 'fraud_illuminating'
]

categorical_columns = [
    'Facebook_Id', 'post_id', 'Page Category', 'Page Admin Top Country', 
    'Post Created', 'Type', 'Video Share Status', 'Is Video Owner?', 'Video Length'
]

print("OVERALL DATASET STATISTICS")
print("=" * 60)

print(f"\nDataset Info:")
print(f"Total Rows: {len(facebook_data)}")
print(f"Total Columns: {len(facebook_data[0].keys()) if facebook_data else 0}")
print(f"Numeric Columns: {len(numeric_columns)}")
print(f"Categorical Columns: {len(categorical_columns)}")

# HELPER FUNCTIONS FOR PURE PYTHON STATISTICS
def get_numeric_values(data, column):
    """Extract numeric values from a column, handling NaN/None"""
    values = []
    for row in data:
        val = row.get(column)
        if val is not None and not (isinstance(val, float) and math.isnan(val)):
            try:
                values.append(float(val))
            except (ValueError, TypeError):
                pass
    return values

def calculate_mean(values):
    """Calculate mean"""
    return sum(values) / len(values) if values else 0

def calculate_std(values):
    """Calculate standard deviation"""
    if len(values) < 2:
        return 0
    mean = calculate_mean(values)
    variance = sum((x - mean) ** 2 for x in values) / len(values)
    return math.sqrt(variance)

def calculate_percentile(sorted_values, percentile):
    """Calculate percentile of sorted values"""
    if not sorted_values:
        return 0
    k = (len(sorted_values) - 1) * percentile / 100
    f = math.floor(k)
    c = math.ceil(k)
    if f == c:
        return sorted_values[int(k)]
    return sorted_values[int(f)] * (c - k) + sorted_values[int(c)] * (k - f)

# NUMERIC STATISTICS
print("\n" + "=" * 60)
print("NUMERIC COLUMNS STATISTICS")
print("=" * 60)

print("\nPure Python numeric statistics (equivalent to .describe()):")
print(f"{'Column':<45} {'count':<8} {'mean':<10} {'std':<10} {'min':<10} {'25%':<10} {'50%':<10} {'75%':<10} {'max':<12}")
print("-" * 150)

numeric_stats_python = {}
for col in numeric_columns:
    values = get_numeric_values(facebook_data, col)
    
    if values:
        sorted_vals = sorted(values)
        count = len(values)
        mean = calculate_mean(values)
        std = calculate_std(values)
        min_val = min(values)
        max_val = max(values)
        q25 = calculate_percentile(sorted_vals, 25)
        q50 = calculate_percentile(sorted_vals, 50)  # median
        q75 = calculate_percentile(sorted_vals, 75)
        
        # Store results
        numeric_stats_python[col] = {
            'count': count,
            'mean': mean,
            'std': std,
            'min': min_val,
            '25%': q25,
            '50%': q50,
            '75%': q75,
            'max': max_val
        }
        
        # Print formatted row with better number formatting
        print(f"{col:<45} {count:<8} {mean:<10.3f} {std:<10.3f} {min_val:<10.3f} {q25:<10.3f} {q50:<10.3f} {q75:<10.3f} {max_val:<12.3f}")
    else:
        print(f"{col:<45} {'No data'}")

print("\nAdditional Numeric Statistics (sum, variance):")
print(f"{'Column':<45} {'sum':<15} {'variance':<15}")
print("-" * 80)

for col in numeric_columns:
    values = get_numeric_values(facebook_data, col)
    if values:
        total_sum = sum(values)
        mean = calculate_mean(values)
        variance = sum((x - mean) ** 2 for x in values) / len(values) if values else 0
        
        # Format large numbers better
        if total_sum >= 1000000:
            sum_formatted = f"{total_sum:,.0f}"
        else:
            sum_formatted = f"{total_sum:.3f}"
            
        print(f"{col:<45} {sum_formatted:<15} {variance:<15.3f}")

# CATEGORICAL STATISTICS
print("\n" + "=" * 60)
print("CATEGORICAL COLUMNS STATISTICS")
print("=" * 60)

categorical_stats_python = {}
for col in categorical_columns:
    print(f"\n--- {col} ---")
    
    # Extract non-null values
    values = []
    for row in facebook_data:
        val = row.get(col)
        if val is not None and val != '':
            values.append(str(val))
    
    count = len(values)
    unique_values = list(set(values))
    unique_count = len(unique_values)
    
    # Most frequent value
    if values:
        counter = Counter(values)
        mode_item = counter.most_common(1)[0]
        mode_value = mode_item[0]
        mode_count = mode_item[1]
    else:
        mode_value = 'N/A'
        mode_count = 0
    
    print(f"Count: {count}")
    print(f"Unique Values: {unique_count}")
    print(f"Most Frequent: '{mode_value}' (appears {mode_count} times)")
    
    # Top 5 values
    print("Top 5 Values:")
    if values:
        counter = Counter(values)
        top_5 = counter.most_common(5)
        for i, (val, count_val) in enumerate(top_5, 1):
            print(f"  {i}. '{val}': {count_val}")
    
    # Store results
    categorical_stats_python[col] = {
        'count': count,
        'unique': unique_count,
        'mode': mode_value,
        'mode_count': mode_count
    }

# End timing
end_time = time.time()
execution_time = end_time - start_time

print("\n" + "=" * 60)
print("EXECUTION SUMMARY")
print("=" * 60)
print(f"Pure Python Part 1 execution time: {execution_time:.4f} seconds")
print(f"Dataset shape: ({len(facebook_data)}, {len(facebook_data[0].keys()) if facebook_data else 0})")
print(f"Numeric columns analyzed: {len(numeric_columns)}")
print(f"Categorical columns analyzed: {len(categorical_columns)}")

# Save results for comparison
python_results_part1 = {
    'numeric_stats': numeric_stats_python,
    'categorical_stats': categorical_stats_python,
    'execution_time': execution_time,
    'dataset_shape': (len(facebook_data), len(facebook_data[0].keys()) if facebook_data else 0)
}

print(f"\nResults saved in 'python_results_part1' variable for comparison")

=== SCRIPT 1 PART 1: PURE PYTHON OVERALL STATISTICS ===
OVERALL DATASET STATISTICS

Dataset Info:
Total Rows: 19009
Total Columns: 52
Numeric Columns: 41
Categorical Columns: 9

NUMERIC COLUMNS STATISTICS

Pure Python numeric statistics (equivalent to .describe()):
Column                                        count    mean       std        min        25%        50%        75%        max         
------------------------------------------------------------------------------------------------------------------------------------------------------
Likes                                         19009    2377.695   11253.174  0.000      31.000     139.000    738.000    351979.000  
Comments                                      19009    901.583    3681.883   0.000      8.000      48.000     354.000    93872.000   
Shares                                        19009    320.539    1722.115   0.000      3.000      21.000     102.000    76150.000   
Love                                          1

# Pandas vs Pure Python Results

## Summary
We tested both Pandas and Pure Python on the same Facebook dataset with 19,009 rows. We found that both methods produce nearly identical results with only tiny differences in standard deviation calculations.

## Key Findings
- **Identical Results**: YES for counts, means, min, max, percentiles, and sums
- **Minor Differences**: Standard deviation differs by less than 0.01% (not important)
- **Speed**: Pandas is 7 times faster (0.093 vs 0.696 seconds)
- **Accuracy**: Both methods handle missing data the same way

## Example Results
For the "Likes" column:
- Count: Both show 19,009 (identical)
- Mean: Both show 2377.695 (identical) 
- Min: Both show 0.000 (identical)
- Max: Both show 351,979.000 (identical)
- Standard deviation: 11253.470 vs 11253.174 (99.997% match)

## Answer: Did We Get Identical Results?
**YES** - The results are identical for all practical purposes. The tiny standard deviation difference is due to different calculation methods (both correct) and doesn't affect any conclusions.

In [12]:
import time
import math
from collections import defaultdict, Counter

# Script 1 Part 2: Pure Python Groupby Analysis for Facebook Dataset
print("=== SCRIPT 1 PART 2: PURE PYTHON GROUPBY ANALYSIS ===")

# Start timing
start_time = time.time()

# Use the same data from Part 1
facebook_data = facebook_posts_clean.to_dict('records')

# Key columns for groupby analysis (same as Pandas script)
engagement_columns = ['Likes', 'Comments', 'Shares', 'Love', 'Wow', 'Haha', 'Sad', 'Angry', 'Care']
view_columns = ['Post Views', 'Total Views', 'Total Interactions']
topic_columns = [
    'covid_topic_illuminating', 'economy_topic_illuminating', 'education_topic_illuminating',
    'environment_topic_illuminating', 'foreign_policy_topic_illuminating', 'governance_topic_illuminating',
    'health_topic_illuminating', 'immigration_topic_illuminating', 'safety_topic_illuminating'
]

# HELPER FUNCTIONS FOR GROUPBY
def get_numeric_value(value):
    """Convert value to float, return None if not possible"""
    if value is None or (isinstance(value, float) and math.isnan(value)):
        return None
    try:
        return float(value)
    except (ValueError, TypeError):
        return None

def calculate_group_stats(values):
    """Calculate count, sum, mean, max for a list of values"""
    clean_values = [v for v in values if v is not None]
    
    if not clean_values:
        return {'count': 0, 'sum': 0, 'mean': 0, 'max': 0}
    
    return {
        'count': len(clean_values),
        'sum': sum(clean_values),
        'mean': sum(clean_values) / len(clean_values),
        'max': max(clean_values)
    }

print("=" * 70)
print("1. GROUPED BY FACEBOOK_ID (Page Analysis)")
print("=" * 70)

# Group data by Facebook_Id
facebook_groups = defaultdict(list)
for row in facebook_data:
    facebook_id = row.get('Facebook_Id')
    if facebook_id:
        facebook_groups[facebook_id].append(row)

print(f"\nFound {len(facebook_groups)} unique Facebook pages")

# Calculate page activity summary
print("\nPage Activity Summary:")
print(f"{'Facebook_Id':<70} {'Posts':<8} {'Likes_Sum':<12} {'Likes_Mean':<12} {'Comments_Sum':<12} {'Comments_Mean':<12}")
print("-" * 130)

page_stats = {}
for facebook_id, posts in facebook_groups.items():
    # Post count
    post_count = len(posts)
    
    # Likes stats
    likes_values = [get_numeric_value(post.get('Likes')) for post in posts]
    likes_stats = calculate_group_stats(likes_values)
    
    # Comments stats
    comments_values = [get_numeric_value(post.get('Comments')) for post in posts]
    comments_stats = calculate_group_stats(comments_values)
    
    # Shares stats
    shares_values = [get_numeric_value(post.get('Shares')) for post in posts]
    shares_stats = calculate_group_stats(shares_values)
    
    # Total Interactions stats
    interactions_values = [get_numeric_value(post.get('Total Interactions')) for post in posts]
    interactions_stats = calculate_group_stats(interactions_values)
    
    # Store results
    page_stats[facebook_id] = {
        'post_count': post_count,
        'likes': likes_stats,
        'comments': comments_stats,
        'shares': shares_stats,
        'interactions': interactions_stats
    }
    
    # Print first 10 pages
    if len(page_stats) <= 10:
        print(f"{facebook_id:<70} {post_count:<8} {likes_stats['sum']:<12.0f} {likes_stats['mean']:<12.3f} {comments_stats['sum']:<12.0f} {comments_stats['mean']:<12.3f}")

# Top 5 pages by different metrics
print("\nTop 5 Pages by Total Likes:")
likes_ranking = sorted(page_stats.items(), key=lambda x: x[1]['likes']['sum'], reverse=True)[:5]
for i, (page_id, stats) in enumerate(likes_ranking, 1):
    print(f"{i}. {page_id[:50]}... : {stats['likes']['sum']:.0f}")

print("\nTop 5 Pages by Total Comments:")
comments_ranking = sorted(page_stats.items(), key=lambda x: x[1]['comments']['sum'], reverse=True)[:5]
for i, (page_id, stats) in enumerate(comments_ranking, 1):
    print(f"{i}. {page_id[:50]}... : {stats['comments']['sum']:.0f}")

print("\nTop 5 Pages by Post Count:")
posts_ranking = sorted(page_stats.items(), key=lambda x: x[1]['post_count'], reverse=True)[:5]
for i, (page_id, stats) in enumerate(posts_ranking, 1):
    print(f"{i}. {page_id[:50]}... : {stats['post_count']}")

# Topic coverage by page
print("\nTopic Coverage by Page (Top 5 pages for each topic):")
for topic in topic_columns[:3]:  # Show first 3 topics
    print(f"\n{topic}:")
    topic_by_page = {}
    
    for facebook_id, posts in facebook_groups.items():
        topic_values = [get_numeric_value(post.get(topic)) for post in posts]
        topic_sum = sum(v for v in topic_values if v is not None)
        topic_by_page[facebook_id] = topic_sum
    
    # Sort and show top 5
    topic_ranking = sorted(topic_by_page.items(), key=lambda x: x[1], reverse=True)[:5]
    for i, (page_id, topic_sum) in enumerate(topic_ranking, 1):
        print(f"  {i}. {page_id[:50]}... : {topic_sum}")

print("\n" + "=" * 70)
print("2. GROUPED BY FACEBOOK_ID + PAGE CATEGORY")
print("=" * 70)

# Group by Facebook_Id AND Page Category
page_category_groups = defaultdict(list)
for row in facebook_data:
    facebook_id = row.get('Facebook_Id', '')
    page_category = row.get('Page Category', 'Unknown')
    
    # Handle NaN values
    if page_category is None or (isinstance(page_category, float) and math.isnan(page_category)):
        page_category = 'Unknown'
    
    key = (facebook_id, str(page_category))
    page_category_groups[key].append(row)

print(f"\nFound {len(page_category_groups)} unique Facebook page + category combinations")

# Calculate page + category stats
print("\nPage + Category Activity Summary (first 15):")
print(f"{'Facebook_Id':<50} {'Category':<20} {'Posts':<8} {'Likes_Sum':<12} {'Likes_Mean':<12}")
print("-" * 110)

page_cat_stats = {}
for (facebook_id, category), posts in list(page_category_groups.items())[:15]:
    post_count = len(posts)
    
    likes_values = [get_numeric_value(post.get('Likes')) for post in posts]
    likes_stats = calculate_group_stats(likes_values)
    
    page_cat_stats[(facebook_id, category)] = {
        'post_count': post_count,
        'likes': likes_stats
    }
    
    print(f"{facebook_id[:50]:<50} {category:<20} {post_count:<8} {likes_stats['sum']:<12.0f} {likes_stats['mean']:<12.3f}")

# Average engagement by Page Category
print("\nAverage Engagement by Page Category:")
category_groups = defaultdict(list)
for row in facebook_data:
    category = row.get('Page Category', 'Unknown')
    if category is None or (isinstance(category, float) and math.isnan(category)):
        category = 'Unknown'
    category_groups[str(category)].append(row)

print(f"{'Category':<25} {'Count':<8} {'Likes_Mean':<12} {'Comments_Mean':<15} {'Shares_Mean':<12}")
print("-" * 80)

category_stats = {}
for category, posts in category_groups.items():
    likes_values = [get_numeric_value(post.get('Likes')) for post in posts]
    comments_values = [get_numeric_value(post.get('Comments')) for post in posts]
    shares_values = [get_numeric_value(post.get('Shares')) for post in posts]
    
    likes_stats = calculate_group_stats(likes_values)
    comments_stats = calculate_group_stats(comments_values)
    shares_stats = calculate_group_stats(shares_values)
    
    category_stats[category] = {
        'count': len(posts),
        'likes_mean': likes_stats['mean'],
        'comments_mean': comments_stats['mean'],
        'shares_mean': shares_stats['mean']
    }
    
    print(f"{category:<25} {len(posts):<8} {likes_stats['mean']:<12.3f} {comments_stats['mean']:<15.3f} {shares_stats['mean']:<12.3f}")

# Post Type Distribution by Page Category
print("\nPost Type Distribution by Page Category:")
type_category_crosstab = defaultdict(lambda: defaultdict(int))

for row in facebook_data:
    category = row.get('Page Category', 'Unknown')
    post_type = row.get('Type', 'Unknown')
    
    if category is None or (isinstance(category, float) and math.isnan(category)):
        category = 'Unknown'
    if post_type is None or (isinstance(post_type, float) and math.isnan(post_type)):
        post_type = 'Unknown'
    
    type_category_crosstab[str(category)][str(post_type)] += 1

# Get all unique post types and categories
categories = sorted(type_category_crosstab.keys())
all_types = set()
for cat_types in type_category_crosstab.values():
    all_types.update(cat_types.keys())
all_types = sorted(all_types)

# Display as a cleaner table
print(f"{'Category':<25}", end="")
for post_type in all_types:
    print(f"{post_type:<12}", end="")
print(" | Total")
print("-" * (25 + 12 * len(all_types) + 8))

for category in categories:
    print(f"{category:<25}", end="")
    row_total = 0
    for post_type in all_types:
        count = type_category_crosstab[category].get(post_type, 0)
        print(f"{count:<12}", end="")
        row_total += count
    print(f" | {row_total}")

# Column totals
print(f"{'TOTAL':<25}", end="")
for post_type in all_types:
    col_total = sum(type_category_crosstab[cat].get(post_type, 0) for cat in categories)
    print(f"{col_total:<12}", end="")
grand_total = sum(sum(cat_types.values()) for cat_types in type_category_crosstab.values())
print(f" | {grand_total}")

# End timing
end_time = time.time()
execution_time = end_time - start_time

print("\n" + "=" * 70)
print("EXECUTION SUMMARY")
print("=" * 70)
print(f"Pure Python Part 2 execution time: {execution_time:.3f} seconds")
print(f"Unique Facebook_Id values: {len(facebook_groups)}")
print(f"Unique Page Categories: {len(category_groups)}")
print(f"Total page-category combinations: {len(page_category_groups)}")

# Save results for comparison
python_results_part2 = {
    'page_stats': page_stats,
    'category_stats': category_stats,
    'execution_time': execution_time,
    'facebook_groups_count': len(facebook_groups),
    'category_groups_count': len(category_groups)
}

print(f"\nResults saved in 'python_results_part2' variable for comparison")

=== SCRIPT 1 PART 2: PURE PYTHON GROUPBY ANALYSIS ===
1. GROUPED BY FACEBOOK_ID (Page Analysis)

Found 21 unique Facebook pages

Page Activity Summary:
Facebook_Id                                                            Posts    Likes_Sum    Likes_Mean   Comments_Sum Comments_Mean
----------------------------------------------------------------------------------------------------------------------------------
a6cb7db6850459b954f7272e14d770617022639f8847cf25e6859b9453a03813       29       1463         50.448       1357         46.793      
7ec2cb4abf8effe3d91de57944c56b938b4f33059a6e33dfe23b549cf229989a       616      24267226     39394.847    8574994      13920.445   
058bd86861262fa71733f8515c34b729d9ebcad3d318e2a133a6c572dd80ba0b       210      12400        59.048       5532         26.343      
ac24f31c4d4b3d5555065fa9558bdca4ab4b5e1379922875ae9e6e32f1d46d25       1231     6207838      5042.923     2268702      1842.975    
7f5731ac8e6959576e781064b5fa76e059200cdfc3f4e0d7b68bc987

**Groupby Analysis Results**

The groupby operations produced identical results across both methods. Top 5 pages by total likes, comments, and post counts matched exactly. Page-level statistics and category-level aggregations showed perfect agreement. All sums, means, and rankings were identical between the two approaches.
Performance Results
For overall statistics, Pandas completed in 0.093 seconds while Pure Python took 0.696 seconds, making Pandas 7.5 times faster. For groupby operations, Pandas finished in 0.074 seconds compared to Pure Python's 0.244 seconds, making Pandas 3.3 times faster for aggregation tasks.

**Key Findings**

Both approaches handle missing data identically and produce the same business insights. Pure Python required significantly more code with manual implementations of statistical functions and groupby logic. Pandas provides built-in optimized functions that are faster and more reliable for production use.
Conclusion
The comparison proves that identical numerical results can be achieved using either approach. While Pure Python helps understand the underlying mathematics, Pandas is clearly superior for practical data analysis due to its speed and ease of use.